src/_artifacts.tf

```diff
@@ -0,0 +1,26 @@
+resource "massdriver_artifact" "api" {
+  field                = "api"
+  provider_resource_id = azurerm_communication_service.main.id
+  name                 = "Azure Communication Services ${azurerm_communication_service.main.name} (${azurerm_communication_service.main.id})"
+  artifact = jsonencode(
+    {
+      data = {
```

Comment:
Maybe it's another pass, but how do we authenticate w/ this service?

Reply:
The `endpoint` (or `hostname` in the `artifacts.tf` file) is set in their function or code, something like this:
```Javascript
const { DefaultAzureCredential } = require("@azure/identity");
const { CommunicationIdentityClient, CommunicationUserToken } = require("@azure/communication-identity");
const { SmsClient, SmsSendRequest } = require("@azure/communication-sms");

const credential = new DefaultAzureCredential();

async function createIdentityAndIssueToken(resourceEndpoint) {
    const client = new CommunicationIdentityClient(resourceEndpoint, credential);
    return await client.createUserAndToken(["chat"]);
}

async function sendSms(resourceEndpoint, fromNumber, toNumber, message) {
    const smsClient = new SmsClient(resourceEndpoint, credential);
    const sendRequest = {
        from: fromNumber,
        to: [toNumber],
        message: message
    };
    return await smsClient.send(
        sendRequest,
        {} //Optional SendOptions
    );
}

async function main() {
    // You can find your endpoint and access key from your resource in the Azure portal
    // e.g. "https://<RESOURCE_NAME>.communication.azure.com";
    const endpoint = "https://<RESOURCE_NAME>.communication.azure.com/"

    
    console.log("Retrieving new Access Token, using Service Principals");
    const result = await createIdentityAndIssueToken(endpoint);
    console.log(`Retrieved Access Token: ${result.token}`);

    console.log("Sending SMS using Service Principals");

    // You will need a phone number from your resource to send an SMS.
    const smsResult = await sendSms(endpoint, "<FROM NUMBER>", "<TO NUMBER>", "Hello from Service Principals");
    console.log(`SMS ID: ${smsResult[0].messageId}`);
    console.log(`Send Result Successful: ${smsResult[0].successful}`);
}

main();
```

Reply:
I'll read more up on that, and this looks like a sweet future app example ^.

My thoughts here are, if I use a service principal, which one? How do I give that one the right permissions to access this service? If the other flow is they go to the endpoint and then log in with their username and password, where does their user get permission for this service?

Reply:
All very good questions. Let me look into this and get back to you. In the meantime, would you mind reviewing the other comments I responded to?

Reply:
Ok so here's the breakdown from what I've been able to find so far:

I've been able to auth by creating an identity w/ Azure AD and then giving it a token. This can be created within the SDK by setting the endpoint as a `const` and also setting [a few environment variables](https://learn.microsoft.com/en-us/azure/communication-services/quickstarts/identity/service-principal-from-cli#set-environment-variables) for an Azure Service Principal using `AZURE_CLIENT_ID`, `AZURE_CLIENT_SECRET`, and `AZURE_TENANT_ID`. I've tested using this method to auth, and it works. I noticed there was another option for an Azure account (human) using `AZURE_USERNAME` and `AZURE_PASSWORD`, but I haven't been able to get that to work (I think because I have 2FA enabled on my account).

Reply:
Do we need to add a 'policies' block to API artdef to support authorization?

Reply:
The above info and link would be great in the operator guide.

@chrisghill @WillBeebe @xpositivityx what do you guys think about adding  `operator.draft.md` that we can keep note sin until a guide is ready to be published.

Reply:
I love this idea.

Reply:
That depends on what developing an app with the azure-communication-service will look like in Massdriver. From the [docs](https://learn.microsoft.com/en-us/azure/communication-services/quickstarts/identity/service-principal-from-cli), they state that service principal credentials will need to be set as env vars, and also that service principal needs the `Contributor` role. From a developer's perspective, what would it look like to them? From my end, I set env vars in my terminal and used the command `node index.js` [to test](https://learn.microsoft.com/en-us/azure/communication-services/quickstarts/identity/service-principal?pivots=programming-language-javascript#run-the-program).

I was thinking the customer would either use the Massdriver service principal, which already has Owner permissions, to auth, _or_ we can create a service principal with `Contributor` for them and output the info they need. The `Contributor` role seems like it needs to be scoped to the subscription from what I can tell in the docs.

Long story short, I don't think the RBAC in the artifact file is needed.

Reply:
> Do we need to add a 'policies' block to API artdef to support authorization?

I think that makes sense here. The docs for Com Service say to use the `DefaultCredential()` method in the SDK. Like @mclacore mentions, it needs Contributor.  

```json
{
  "data": {
    "type": "object",
    "required": [
      "api"
    ],
    "properties": {
      "etc": {},
      "api": {},
      "security": {
        "title": "Security details",
        "type": "object",
        "anyOf": [
          {
            "$ref": "../types/aws-security.json"
          },
          {
            "$ref": "../types/azure-security.json"
          },
          {
            "$ref": "../types/gcp-security.json"
          }
        ]
      }
    }
  }
}
```

It would be nice is there was _any role_ that was more restrictive than the Contributor role. 

```hcl
  data_security = {
    iam = {
      "read/write" = {
        role  = "Contributor"
        scope = azurerm_communication_service.main.id
      },
    }
  }
```

Reply:
If that's too much cloud for the api artifact definition we can make an artifact definition just for the azure com service.

Reply:
I'm pretty sure that would need to be scoped to the subscription: https://learn.microsoft.com/en-us/azure/communication-services/quickstarts/identity/service-principal-from-cli#creating-an-azure-active-directory-registered-application

So it would probably look like:
```
  data_security = {
    iam = {
      "read/write" = {
        role  = "Contributor"
        scope = "/subscriptions/${var.azure_service_principal.data.subscription_id}"
      },
    }
  }
```

Reply:
We definitely don't want to run user apps using the SP they give us during onboarding. Apps should be tightly scoped using Auto IAM so they have principle of least priv.

I wouldn't worry about human user auth, this is something we have dismissed supporting today. We have a grander scope to add support for that later this year after RBAC.

They way this _should_ work is:

1. the bundle provisions the resource and outputs the IAM policies/scopes/etc
2. mdxc creates an identity (role, principal, whatev) for the app and selects the policy/scope to bind.

Reply:
Ok cool. I'm understanding now. Based on the [docs](https://learn.microsoft.com/en-us/azure/communication-services/quickstarts/identity/service-principal-from-cli#creating-an-azure-active-directory-registered-application), we can add a policies block based on Beebe's example.

Back to your original question:
> Do we need to add a 'policies' block to API artdef to support authorization?

I think we do. What would be the best approach for that?

Reply:
I tested auth to comm service SDK using a service principal with `Contributor` role to the comm service itself only, not to the subscription, and I was issued a token. So I think scoping to the comm service only is the best way, even though the docs scope to the subscription.

Reply:
> This can be created within the SDK by setting the endpoint as a const and also setting [a few environment variables](https://learn.microsoft.com/en-us/azure/communication-services/quickstarts/identity/service-principal-from-cli#set-environment-variables) for an Azure Service Principal using `AZURE_CLIENT_ID`, `AZURE_CLIENT_SECRET`, and `AZURE_TENANT_ID`. I've tested using this method to auth, and it works. I noticed there was another option for an Azure account (human) using `AZURE_USERNAME` and `AZURE_PASSWORD`

I'm curious if this works for managed identities (system assigned and user assigned). It seems we are trying to move that direction since its considered "best-practice" by Azure and is credential-less.

Reply:
Identity SDK doc for ref: https://azuresdkdocs.blob.core.windows.net/$web/javascript/azure-communication-identity/1.0.0/index.html

Reply:
So what would be the `role` and `scope` we would be emitting in the security block of the API artifact? role is `Contributor`? Not sure I follow the scope... is that the subscription?

Reply:
The scope will be to the `azurerm_communication_service.main.id`. I tested it locally on my end, and I was able to auth and generate a token with a service principal scoped to the communication service only.